cpp/src/arrow/compute/exec/expression.h

```diff
@@ -91,6 +91,9 @@ class ARROW_EXPORT Expression {
   /// Return true if this expression could evaluate to true.
   bool IsSatisfiable() const;
 
+  /// Return true if this expression has no clauses.
+  bool IsEmpty() const;
```

Comment:
Thinking about it more, I'm not a big fan of this `literal(true)` convention. Currently, null expressions are allowed (by calling the default constructor). It's very bug-prone leaving code around that assumes expressions are non-null while having no good way of checking this. Maybe instead of introducing `IsEmpty` you can do `== Expression()`, but at that point I feel `IsEmpty` is cleaner. 

But anyway, If this is the convention, then the default constructor should be deleted to eliminate the possibility of there ever being a null `Expression`.

Reply:
`Expression()` is currently allowed to enable preallocation in `std::vector<Expression>`. IIUC, the only circumstances under which an expression would be empty are (a) a preallocated element of a container which will be overwritten shortly (b) a moved-from variable, *any* use of which should be avoided or (c) when used as a special-case placeholder for non-empty expressions.

If I'm missing something, please file a follow up JIRA for adding `IsEmpty()`.

Reply:
Well I think it's more of just a gotcha kind of thing. You're allowed to create an object that's in an invalid state, and the rest of the codebase assumes it's not in that state. My thinking is that if you're allowed to create an object in that state, it should be considered a valid value, and should therefore be checked against.

Reply:
If it were possible to forbid default construction of `Expression` without breaking `vector<Expression>::resize` I would do so.  As I've been trying to express, it's not really valid for humans to write `Expression defaulted;` but there's no attribute or pragma I'm aware of to make the compiler warn about that. Maybe the solution is to add a large warning comment for `Expression() = default;`

Reply:
Right, so I'd argue the solution is to make `Expression defaulted;` valid. Anyway, this is probably out of scope for this PR.